shotover-proxy/src/transforms/cassandra/sink_cluster.rs

```diff
@@ -93,60 +123,185 @@ impl CassandraSinkCluster {
-                                {
-                                    self.failed_requests.increment(1);
+        // Attempt to populate nodes list if we still dont have one yet
+        if self.nodes.is_empty() {
+            let nodes_shared = self.nodes_shared.read().await;
+            self.nodes.extend(nodes_shared.iter().cloned());
+        }
+
+        let system_peers_index = self.get_system_peers_index(&mut messages);
+
+        for message in &messages {
+            if let Some(last) = self.contact_point_handshake.last_mut() {
+                if let Some(Frame::Cassandra(CassandraFrame {
+                    operation: CassandraOperation::AuthResponse(_),
+                    ..
+                })) = last.frame()
+                {
+                    self.contact_point_handshake_complete = true;
+                    break;
+                }
+            }
+            self.contact_point_handshake.push(message.clone());
+        }
+
+        let mut responses_future = FuturesOrdered::new();
+        for message in messages {
+            let (return_chan_tx, return_chan_rx) = oneshot::channel();
+            if self.nodes.is_empty() || !self.contact_point_handshake_complete {
+                self.contact_point_connection.as_mut().unwrap()
+            } else {
+                let random_node = self.nodes.choose_mut(&mut self.rng).unwrap();
+                random_node
+                    .get_connection(
+                        &self.contact_point_handshake,
+                        &self.tls,
+                        &self.pushed_messages_tx,
+                    )
+                    .await?
+            }
+            .send(message, return_chan_tx)?;
+
+            responses_future.push(return_chan_rx)
+        }
+
+        let mut responses =
+            super::connection::receive(&self.failed_requests, responses_future).await?;
+
+        if let Some(query_occured) = system_peers_index {
+            if let Some(response) = responses.get_mut(query_occured.message_index) {
+                let new_nodes = get_nodes_from_system_peers(
+                    response,
+                    query_occured.alias_to_name,
+                    &self.data_center,
+                );
+                if new_nodes.len() > 1 {
+                    self.nodes.clear();
+                    self.nodes.extend(new_nodes.iter().cloned());
+
+                    let mut write_lock = self.nodes_shared.write().await;
+                    let expensive_drop = std::mem::replace(&mut *write_lock, new_nodes);
+
+                    // Make sure to drop write_lock before the expensive_drop which may have to perform many deallocations.
+                    std::mem::drop(write_lock);
+                    std::mem::drop(expensive_drop);
+                }
+            }
+        }
+
+        Ok(responses)
+    }
+
+    fn get_system_peers_index(&self, messages: &mut [Message]) -> Option<QueryOccured> {
```

Comment:
Can we combine this with the similar function in the peers rewrite transform?

Reply:
I deleted the function instead ;)